src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs

```diff
@@ -0,0 +1,479 @@
+
+            ArgumentException.ThrowIfNullOrEmpty(organizationalUnitName);
+            EncodeComponent(Oids.OrganizationalUnit, organizationalUnitName, UniversalTagNumber.UTF8String);
+        }
+
+        /// <summary>
+        /// Adds a state or province name attribute.
+        /// </summary>
+        /// <param name="stateOrProvinceName">The state or province name to add.</param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="stateOrProvinceName" /> is <see langword="null" />.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="stateOrProvinceName" /> is empty.
+        /// </exception>
+        /// <remarks>
+        /// This encodes an attribute with the OID 2.5.4.8 as a UTF8String.
+        /// </remarks>
+        public void AddStateOrProvinceName(string stateOrProvinceName)
+        {
+            // ITU T-REC X.520 Annex A:
+            // id-at-stateOrProvinceName
+            // WITH SYNTAX UnboundedDirectoryString
+
+            ArgumentException.ThrowIfNullOrEmpty(stateOrProvinceName);
+            EncodeComponent(Oids.StateOrProvinceName, stateOrProvinceName, UniversalTagNumber.UTF8String);
+        }
+
+        /// <summary>
+        /// Adds a domain component attribute.
+        /// </summary>
+        /// <param name="domainComponent">The domain component to add.</param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="domainComponent" /> is <see langword="null" />.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="domainComponent" /> is empty.
+        /// </exception>
+        /// <remarks>
+        /// This encodes an attribute with the OID 0.9.2342.19200300.100.1.25 as an IA5String.
+        /// </remarks>
+        public void AddDomainComponent(string domainComponent)
+        {
+            // RFC 5912
+            // id-domainComponent      AttributeType ::=
+            //  { itu-t(0) data(9) pss(2342) ucl(19200300) pilot(100)
+            //  pilotAttributeType(1) 25 }
+            //  at-domainComponent ATTRIBUTE ::= {TYPE IA5String
+            //  IDENTIFIED BY id-domainComponent }
+
+            ArgumentException.ThrowIfNullOrEmpty(domainComponent);
+            EncodeComponent(Oids.DomainComponent, domainComponent, UniversalTagNumber.IA5String);
+        }
+
+        /// <summary>
+        /// Builds an <see cref="X500DistinguishedName" /> that represents the encoded attributes.
+        /// </summary>
+        /// <returns>
+        /// An <see cref="X500DistinguishedName" /> that represents the encoded attributes.
+        /// </returns>
+        public X500DistinguishedName Build()
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+
+            using (writer.PushSequence())
+            {
+                foreach (byte[] component in _encodedComponents)
```

Comment:
I feel like this loop is backwards from what we wanted it to be.  Because strings and X500DN reversing and yadda yadda yadda.